### ISSUE TITLE
Lookup operator on arrays of maps

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qtspecs#115]
I've been converting the XMark data files and queries from XML to JSON.

Here's part of Q20 in its XML form:

```
 <preferred>
  {count (/site/people/person/profile[@income >= 100000])}
 </preferred>
```

which becomes this, when we access the JSON form of the data:

```
 <preferred>
  {count (?people?*?profile[?income >= 100000])}
 </preferred>
```

It can get worse, for example Q16 has

`exists ($a?annotation?description?parlist?*?parlist?*?text?*?emph?*?keyword?*?("§"))
`

These paths arise because a structured derived from JSON often includes map entries whose values are arrays.

It's very hard to get these paths right, and it's hard to produce good diagnostics when you get them wrong.

I'd like to allow the `?*?` "operators" to be replaced with a simple "?". This isn't difficult. Currently the rules for the lookup operator say:

```
If the context item is an array:
If the [KeySpecifier] is an NCName, the [UnaryLookup](https://www.w3.org/TR/xpath-31/#doc-xpath31-UnaryLookup) operator raises a type error 
```

All that's needed is to change this to say that if the context item is an array, and the KeySpecifier is an NCName, then the array must be an array of maps and the lookup is applied to these maps.

**Comment by ChristianGruen** [comment]
Interesting tweak. The majority of the input can be expected to be regular, but I’m trying to understand what we get if we have mixed types:

```xquery
let $map := map { 1: 1984, 'key': 'value' }
let $seq := ($map, [ $map ])
return $seq?(1, 'key')
```

Would this give us `1984`, `map { 1: 1984, 'key': 'value' }`, `value`, `value`,  as result?

**Comment by michaelhkay** [comment]
I'm open to suggestions on that. I think the best analogy with the current rules would be that the operator gets applied to EVERY item in the array and it's a type error if any of them fails.

**Comment by ChristianGruen** [comment]
Sounds reasonable. We may additionally need to clarify what happens if arrays are nested:

```xquery
[ [ [ map { 'x': 'y' } ] ] ]?x   → y?
```

**Comment by ChristianGruen** [comment]
I propose to close this issue, now that we have a more general approach with deep lookups.

**Comment by ndw** [comment]
The CG agreed to close this issue without further action at [meeting 073](https://qt4cg.org/meeting/minutes/2024/04-16.html).

**Comment by michaelhkay** [comment]
I'm re-opening this in the light of the case study I'm conducting over at issue #1786.

When we convert XML structures to JSON, we often end up with arrays of singleton maps, such as:

```
{ "digest":[
    {
      "package": "net.sf.saxon.tree",
      "class": [
        { "name":"NamespaceNode" },
        { "implements":{ "name":"net.sf.saxon.om.NodeInfo" } },
        { "constructor":{ "params":"net.sf.saxon.om.NodeInfo|net.sf.saxon.om.NamespaceBinding|int" } },
        { "field":{ "name":"element", "type":"net.sf.saxon.om.NodeInfo" } },
        { "field":{ "name":"nsBinding", "type":"net.sf.saxon.om.NamespaceBinding" } },
        { "field":{ "name":"position", "type":"int" } },
        { "field":{ "name":"fingerprint", "type":"int" } },
        { "method":{ "name":"getTreeInfo", "returns":"net.sf.saxon.om.TreeInfo" } },
        { "method":{ "name":"head", "returns":"net.sf.saxon.om.NodeInfo", "csReturns":"net.sf.saxon.om.Item" } },
        { "method":{ "name":"getNodeKind", "returns":"int" } },
        { "method":{ "name":"equals", "returns":"boolean", "sig":"java.lang.Object", "params":"java.lang.Object" } },
        { "method":{ "name":"hashCode", "returns":"int" } },
        { "method":{ "name":"getSystemId", "returns":"java.lang.String" } },
        { "method":{ "name":"getPublicId", "returns":"java.lang.String" } },
        { "method":{ "name":"getBaseURI", "returns":"java.lang.String" } },
        { "method":{ "name":"getLineNumber", "returns":"int" } },
        { "method":{ "name":"getColumnNumber", "returns":"int" } },
        ....
] } ] }
```
When $X a sequence of maps, and expression such as $X?name works perfectly well. But when $X is an array of maps (such as the value associated with the key "class" in this example), it doesn't: we have to write `$X?*?name`. We can't cut corners by writing `$X??name` because this would over-retrieve.

Currently, when $X is an array, `$X?name` (where name is an NCName or a string literal) is a type error. I propose that it should instead be interpreted as `?X?*?name`.

For example, given the above structure, I can now write

```
  ?digest [?package="net.sf.saxon.tree"] 
  ?class [?name="NamespaceNode"] 
  ?method [?name="getTreeInfo"] 
  ?returns 

```
in place of

```
 ?digest [?*?package="net.sf.saxon.tree"]
  ?*?class[?*?name="NamespaceNode"]
  ?*?method[?name="getTreeInfo"]
  ?returns 
```

which feels an awful lot easier to get right.

**Comment by ChristianGruen** [comment]
> Currently, when $X is an array, `$X?name` (where name is an NCName or a string literal) is a type error. I propose that it should instead be interpreted as `$X?*?name`.

This convenience shortcut would be based on the assumption that most (often JSON-based) maps use strings as keys – but it could cause confusion with maps that have integer keys that are not too rare in practice.

For the following array, `?1` would return a map, while `?one` would return an integer:

```xquery
[
  { 1: 'one' },
  { 'one': 1 }
]
```

Maybe this is confusing? Maybe the existing Saturday (even if it requires two more characters) is more self-explanatory?

**Comment by michaelhkay** [comment]
>Maybe this is confusing?

Yes, I'm aware of the risk. Having spent a couple of days debugging this kind of code though, and getting myself extremely confused by the current situation, I think on balance that the change would be beneficial.

**Comment by ChristianGruen** [comment]
The risk may be low indeed. In practice, maps with integer keys are mostly dictionaries that are being looked up dynamically. I have checked our code base for `?*?` (with optional whitespace): In all cases, the characters were followed by either strings or parentheses (i.e., no integers).

> Currently, when $X is an array, `$X?name` (where name is an NCName or a string literal) is a type error. I propose that it should instead be interpreted as `$X?*?name`.

By all means, I think the decision should be made at runtime, to treat `$map-with-arrays?$key` and `$map-with-arrays?('key1', 'key2')` equally.

It could be worth mentioning somewhere that…

* `$map-with-arrays?key` may result in multiple sequence-concatenated results (which may be fine, as it is in line with `$map?*`),
* `$map-with-arrays?fn()` may result in multiple method calls, and
* `$map?$key`, `$map($key)`, and `map:get($map, $key)` are not equivalent anymore (which is already the case, considering the special treatment of methods).

**Comment by ruv** [comment]
> I would welcome the addition of this tweak to JNode steps

I don't sure that such a tweak is needed.

<details>
<summary>
Discussed structure
</summary>

(a copy from the above [message](https://github.com/qt4cg/qtspecs/issues/115#issuecomment-2904038013))

```json
{ "digest":[
    {
      "package": "net.sf.saxon.tree",
      "class": [
        { "name":"NamespaceNode" },
        { "implements":{ "name":"net.sf.saxon.om.NodeInfo" } },
        { "constructor":{ "params":"net.sf.saxon.om.NodeInfo|net.sf.saxon.om.NamespaceBinding|int" } },
        { "field":{ "name":"element", "type":"net.sf.saxon.om.NodeInfo" } },
        { "field":{ "name":"nsBinding", "type":"net.sf.saxon.om.NamespaceBinding" } },
        { "field":{ "name":"position", "type":"int" } },
        { "field":{ "name":"fingerprint", "type":"int" } },
        { "method":{ "name":"getTreeInfo", "returns":"net.sf.saxon.om.TreeInfo" } },
        { "method":{ "name":"head", "returns":"net.sf.saxon.om.NodeInfo", "csReturns":"net.sf.saxon.om.Item" } },
        { "method":{ "name":"getNodeKind", "returns":"int" } },
        { "method":{ "name":"equals", "returns":"boolean", "sig":"java.lang.Object", "params":"java.lang.Object" } },
        { "method":{ "name":"hashCode", "returns":"int" } },
        { "method":{ "name":"getSystemId", "returns":"java.lang.String" } },
        { "method":{ "name":"getPublicId", "returns":"java.lang.String" } },
        { "method":{ "name":"getBaseURI", "returns":"java.lang.String" } },
        { "method":{ "name":"getLineNumber", "returns":"int" } },
        { "method":{ "name":"getColumnNumber", "returns":"int" } },
        ....
] } ] }
```
</details>

A *tweaked* path:
```xquery
  ?digest [?package="net.sf.saxon.tree"] 
  ?class [?name="NamespaceNode"] 
  ?method [?name="getTreeInfo"] 
  ?returns 
```
In this tweaked path it is unclear do we want to select `class`
  - from **all** the members of specific `digest`:
    - `digest[ */package="net.sf.saxon.tree"]/*`
  - from **only** the specific members of `digest`:
    - `digest/*[ package="net.sf.saxon.tree"]`


An XPath with support of JNode:
```xquery
  digest/ * [ package="net.sf.saxon.tree"]/
  class [ */name="NamespaceNode"]/
  */method [ name="getTreeInfo"]/
  returns
```

This path reflects the structure very well:
  - select members of `digest` that have an entry `package="net.sf.saxon.tree"`
  - select `class` entries which have a member with entry `name="NamespaceNode"`
  - for all members, select `method` entries that have entry `name="getTreeInfo"`
  - select `returns` entries (which are siblings of `name` in the above line)

**Comment by michaelhkay** [comment]
I think that with the flexibiity we now have with JNode-based path expressions. we can close this issue with no action.
